### PR TITLE
New EVM opcode 'TERMINATE'

### PR DESCRIPTION
**Pull request by Arachnid**

### CODE REVIEW

**Comment by alex-miller-0** [comment]
This might be difficult to test.

**Comment by veox** [comment]
**TL;DR:** Proposed `TERMINATE` opcode opens a "multiversal extortion" vulnerability.

Suggest closing preemptively, before being put on any track, so as not to waste any further resources, across the whole set of currently available multiverses.

-----

### Proposal does not describe "universe", or "destroy the universe"

First, proposed opcode `0xFC` [has not been designated AFAIK](https://github.com/ethereum/yellowpaper/blob/6878240048d49a4ddbecf9ade7fbae6cbc3bac8d/Paper.tex#L1997) (link to `0xF4` in `Paper.tex` at current `master` branch `HEAD` in `ethereum/yellowpaper` repo), so can indeed be considered.

However, the proposal does not sufficiently describe operation "destroy the universe" - i.e. the operation's state transition function.

Further discussion is therefore impeded, but I will try my best.

Tentatively, it can be described as **an operation that forbids any operation other than itself**. Following a `TERMINATE` operation, the only valid subsequent operation would then be `TERMINATE`.

How a conforming EVM chooses to achieve this would, of course, be left to that specific impl-n:

* for hardware impl-s, the approach may be as simple as setting "current operation" register and "address counter" register read-only, in effect discarding any further writes to said memory registers;
* for software impl-s, the approach depends on the abstraction level used: e.g. for Assembly impl-s, `terminate: goto terminate` may suffice; for Python, a `next_op_generator` may need to be short-circuited to return `current_op`; etc. (sadly, I'm not versed enough in current EVMs to propose viable impl-n details).

A node running such an EVM could safely stop all external interaction upon encountering `TERMINATE`. In effect, it would signal "final computation state", a.k.a. "no input necessary", allowing the remaining nodes not to waste network bandwidth trying to reach it.

### Multiversal extortion attack

AFAIU from the proposal's [Motivation section](https://github.com/Arachnid/EIPs/blob/73ca1777a0d8b0e3255445796033709e5db6ca9d/EIPS/eip-draft-terminate.md#motivation), the intended use is achieving improved computational complexity for certain algorithms, by preemptively defining desired outcomes.

IMO this can be seen as a logical first step in harnessing the quantum computation capabilities of the EVM, but it also opens a certain attack vector that **can be executed across all multiverses that implement the proposal**.

Note first that with the (tentative!) `TERMINATE` transition function described above, only conformant EVM impl-s (and by extension their universes) will be impacted. I.e. the effect of executing `TERMINATE` is limited to a subset of multiverse where _both_ said `TERMINATE` got executed, _and_ the impl-n was conformant.

Note also that it is possible to pre-deploy a contract that:

* would have its bytecode read as `NOOP` ATM, for all current EVMs; but
* would read as `TERMINATE` for all EVMs conforming to the proposal in the future.

(If all current `NOOP`s were changed to `INVALID` first, then it would still be possible to pre-deploy such a contract. Even if contract validation during contract creation was enforced first, it would still be possible to deploy such a contract for all universe+EVM sets that implemented the proposal (i.e. that changed given opcode from `INVALID` to `TERMINATE`), and ignoring contract creation failure for all universes that didn't implement it.)

The contract can be constructed such that it **would check whether the origin account** (of a transaction) **has any balance, and call `TERMINATE` if so**.

(AFAIK, such checks can't be prevented - see opcodes `0x31` (`BALANCE`), `0x32` (`ORIGIN`), `0x33` (`CALLER`) in [yellowpaper](http://yellowpaper.io/).)

This would **force any calls** to such an "extortion" contract (by rational actors) **to include all origin account balance** - otherwise running the risk of a `TERMINATE`. 

There would still, of course, be cases where calls to the contract were made by accident, or as a result of off-chain extortion.

Therefore, for the subset of the multiverse that implemented such an opcode, the only stable non-terminated universe would be that which can guarantee all balance being in the extortion contract.

The above describes the state of equilibrium such a system would gravitate towards, - although, mind you, the _paths_ toward such a state are innumerable.

### `TERMINATE++`

It can be argued that in such a universe, balance changes to all accounts other than the extortion contract's would be transitive. That is, they would not persist longer than the duration of a transaction (which from the chain's PoV is atomic), and would need to guard against inadvertently calling the extortion contract while having balance associated.

It wouldn't be outlandish to speculate that, if PoW or PoS rules were still in effect, all "reward" would be forwarded to the extortion account, so as not to inadvertently trigger termination of the (given) universe.

Almost all of these transactions would probably be initiated by the extortion contract itself (and its associated account), corresponding roughly to the [heat death of the (given) universe](https://en.wikipedia.org/wiki/Heat_death_of_the_universe).

However, given infinite time, the balance of the extortion contract's account could itself happen to be zero at a particular moment of time, - due to integer overflow, floating point rounding error, or any other possible limitation of representing account balance. If the extortion account were to make a call to itself at such a moment, it would result in a `TERMINATE`.

-----

P.S. These are my thoughts. They have been proof-read by a third party that (so far, to the best of my knowledge) wishes to remain anonymous. This note has been added at their request.

**Comment by veox** [comment]
P.P.S. I haven't checked (yet) if including all the balance of an account in a transaction as gas would make `BALANCE` of that account's address read as zero, but that was my assumption while writing the "TERMINATE++" section.

If the above is not true - well, then I've just jolly-cheated myself out of a bounty payment. :(

**Comment by wanderer** [comment]
this will go great with [EVM 3.0](https://www.youtube.com/watch?v=uNjxe8ShM-8)!

**Comment by Arachnid** [comment]
I'm closing this as suggested, due to unacceptable risk of unintended side-effects.

**Comment by gcolvin** [comment]
@wanderer Yes, much better to destroy the universe than to program in Powerpoint.
